src/traits.rs

```diff
@@ -237,59 +236,46 @@ where
         <D as DistAlgorithm>::NodeId: 'i,
     {
         let messages = &mut self.messages;
-        let (mut passed_msgs, failed_msgs): (Vec<_>, Vec<_>) =
-            messages
-                .drain(..)
-                .partition(|TargetedMessage { target, message }| match target {
+        let pass =
+            |TargetedMessage { target, message }: &TargetedMessage<D::Message, D::NodeId>| {
+                match target {
                     Target::All => peer_epochs
                         .values()
                         .all(|&them| message.is_accepted(them, max_future_epochs)),
                     Target::Node(id) => peer_epochs
                         .get(&id)
                         .map_or(false, |&them| message.is_accepted(them, max_future_epochs)),
-                });
+                }
+            };
         // `Target::All` messages contained in the result of the partitioning are analyzed further
         // and each split into two sets of point messages: those which can be sent without delay and
         // those which should be postponed.
-        let remote_nodes: BTreeSet<&D::NodeId> = peer_epochs.keys().collect();
         let mut deferred_msgs: Vec<(D::NodeId, D::Message)> = Vec::new();
-        for msg in failed_msgs {
-            let m = msg.message;
-            match msg.target {
-                Target::Node(id) => {
-                    let defer = {
-                        let lagging = |&them| {
-                            !(m.is_accepted(them, max_future_epochs) || m.is_obsolete(them))
+        let mut passed_msgs: Vec<_> = Vec::new();
+        for msg in messages.drain(..) {
+            if pass(&msg) {
+                passed_msgs.push(msg);
+            } else {
+                let m = msg.message;
+                match msg.target {
+                    Target::Node(ref id) => {
+                        let defer = {
+                            let lagging = |&them| {
+                                !(m.is_accepted(them, max_future_epochs) || m.is_obsolete(them))
+                            };
+                            peer_epochs.get(&id).map_or(true, lagging)
```

Comment:
Now that `peer_epochs` contains all peers as keys, we can probably drop the message if the peer isn't found?

And then we already know at this point that `!m.is_accepted(them, max_future_epochs)`. I think the code would look easier if `pass` was fully merged into this `match` statement here, instead of matching the same message twice. Then it would just say something like this here:

```rust
if let Some(them) = peer_epochs.get(&id) {
    if m.is_accepted(them, max_future_epochs) {
        passed_msgs.push(msg);
    } else if !m.is_obsolete(them) {
        deferred_msgs.push(msg);
    }
}
```

Reply:
You wouldn't be able to push `msg` without cloning since in that case it is borrowed by the match statement.

Reply:
But we can instead clone the target of the message only, which is cheaper compared to cloning the whole message, even though it will be done on every outgoing message.